### PR TITLE
2753 grant consent command

### PR DESCRIPTION
**Pull request by ckr123**

### CODE REVIEW

File: domains/authorization/Authorization.API/API.IntegrationTests/API.IntegrationTests/Setup/DatabaseTest.cs [hunk 1]
```diff
@@ -1,25 +1,39 @@
...
+        var dbInfo = fixture.PostgresContainer.CreateNewDatabase().Result;
 
-        _scope = factory.Services.CreateScope();
+        var options = new DbContextOptionsBuilder<ApplicationDbContext>().UseNpgsql(dbInfo.ConnectionString).Options;
+        using var dbContext = new ApplicationDbContext(options);
+        dbContext.Database.Migrate();
+
+        using var dbConnection = new NpgsqlConnection(dbInfo.ConnectionString);
+        dbConnection.Open();
+        var respawner = Respawner.CreateAsync(dbConnection, new RespawnerOptions
```
Comment (TopSwagCode) [inline domains/authorization/Authorization.API/API.IntegrationTests/API.IntegrationTests/Setup/DatabaseTest.cs:28]: What is respawner ? :D

File: domains/authorization/Authorization.API/API.IntegrationTests/API.IntegrationTests/Setup/IntegrationTestFixture.cs [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    public PostgresContainer PostgresContainer { get; } = new();
+
+    public async Task InitializeAsync()
+    {
+        await PostgresContainer.InitializeAsync();
+
+        using var scope = Services.CreateScope();
+        await using var db = scope.ServiceProvider.GetRequiredService<ApplicationDbContext>();
+        await using var connection = db.Database.GetDbConnection();
+        await connection.OpenAsync();
```
Comment (TopSwagCode) [inline domains/authorization/Authorization.API/API.IntegrationTests/API.IntegrationTests/Setup/IntegrationTestFixture.cs:40]: Why OpenAsync?

File: domains/authorization/Authorization.API/API.IntegrationTests/API.IntegrationTests/Setup/IntegrationTestFixture.cs [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    public static void EnsureDbCreated<T>(this IServiceCollection services) where T : DbContext
+    {
+        using var scope = services.BuildServiceProvider().CreateScope();
+        var serviceProvider = scope.ServiceProvider;
+        var context = serviceProvider.GetRequiredService<T>();
+        context.Database.EnsureCreated();
+    }
+}
+
+public class TestAuthHandler : AuthenticationHandler<AuthenticationSchemeOptions>
```
Comment (TopSwagCode) [inline domains/authorization/Authorization.API/API.IntegrationTests/API.IntegrationTests/Setup/IntegrationTestFixture.cs:138]: Can we move TestAuthHandler to own file? :)

File: domains/authorization/Authorization.API/API.UnitTests/API.UnitTests/Repository/FakeAffilicationRepository.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+using API.Repository;
+using MockQueryable.NSubstitute;
+
+namespace API.UnitTests.Repository;
+
+public class FakeAffiliationRepository : IAffiliationRepository
+{
+    private readonly List<Affiliation> _entities = [];
+
+    public Task<Affiliation> GetAsync(object[] keys, CancellationToken cancellationToken)
```
Comment (TopSwagCode) [inline domains/authorization/Authorization.API/API.UnitTests/API.UnitTests/Repository/FakeAffilicationRepository.cs:12]: I dont entirely understand: object[] keys. Is it a combined key?

File: domains/authorization/Authorization.API/API.UnitTests/API.UnitTests/Repository/FakeAffilicationRepository.cs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        if (entity is null)
+        {
+            throw new EntityNotFoundException(organisationId + userId.ToString(), nameof(Affiliation));
+        }
+        return Task.FromResult(entity);
+    }
+
+    public Task<Affiliation> GetAsync(Guid id, CancellationToken cancellationToken)
+    {
+        throw new NotImplementedException();
```
Comment (TopSwagCode) [inline domains/authorization/Authorization.API/API.UnitTests/API.UnitTests/Repository/FakeAffilicationRepository.cs:27]: This perhaps shows that generic repository, doesn't make sense?

File: domains/authorization/Authorization.API/API.UnitTests/API.UnitTests/Repository/FakeRepository.cs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+namespace API.UnitTests.Repository;
+
+public class FakeGenericRepository<TEntity> : IGenericRepository<TEntity> where TEntity : class, IEntity<Guid>
+{
+    private readonly List<TEntity> _entities = new List<TEntity>();
+
+    public Task<TEntity> GetAsync(object[] keys, CancellationToken cancellationToken)
+    {
+        throw new NotImplementedException();
```
Comment (TopSwagCode) [inline domains/authorization/Authorization.API/API.UnitTests/API.UnitTests/Repository/FakeRepository.cs:14]: generic repository, doesn't make sense?

File: domains/authorization/Authorization.API/API/Authorization/[Features]/GetConsentForClientQuery.cs [hunk 7]
```diff
@@ -1,30 +1,39 @@
...
-        return Task.FromResult(result);
+        return result;
     }
 }
 
-public record GetConsentForClientQuery(string ClientId) : IRequest<GetConsentForClientQueryResult>;
+public record GetConsentForClientQuery(Guid ClientId) : IRequest<GetConsentForClientQueryResult>;
 
-public record GetConsentForClientQueryResult(string Sub, string Name, string SubType, string OrgName, IEnumerable<string> OrgIds, string Scope);
+public record GetConsentForClientQueryResult(Guid Sub, string SubType, string OrgName, IEnumerable<string> OrgIds, string Scope);
```
Comment (TopSwagCode) [inline domains/authorization/Authorization.API/API/Authorization/[Features]/GetConsentForClientQuery.cs:39]: IEnumerable<string> OrgIds -> IEnumerable<Guid> OrgIds

File: domains/authorization/Authorization.API/API/Authorization/[Features]/GetConsentForUserQuery.cs [hunk 8]
```diff
@@ -15,6 +15,6 @@ public Task<GetConsentForUserQueryResult> Handle(GetConsentForUserQuery query, C
     }
 }
 
-public record GetConsentForUserQuery(string Sub, string Name, string OrgName, string OrgCvr) : IRequest<GetConsentForUserQueryResult>;
+public record GetConsentForUserQuery(Guid Sub, string Name, string OrgName, string OrgCvr) : IRequest<GetConsentForUserQueryResult>;
 
-public record GetConsentForUserQueryResult(string Sub, string Name, string SubType, string OrgName, IEnumerable<string> OrgIds, string Scope);
```
Comment (TopSwagCode) [inline domains/authorization/Authorization.API/API/Authorization/[Features]/GetConsentForUserQuery.cs:20]: IEnumerable<Guid> OrgIds

File: domains/authorization/Authorization.API/API/Authorization/[Features]/GrantConsentCommand.cs [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+            .Where(a => a.UserId == command.userId && a.OrganizationId == command.organizationId)
+            .FirstOrDefaultAsync(cancellationToken);
+
+        if (affiliation is null)
+        {
+            throw new UserNotAffiliatedWithOrganizationCommandException();
+        }
+
+        var client = clientRepository.Query()
+                         .AsEnumerable()
```
Comment (TopSwagCode) [inline domains/authorization/Authorization.API/API/Authorization/[Features]/GrantConsentCommand.cs:39]: .AsEnumerable() <--- I am afraid this will get entire database? Can we just use first or default?

File: domains/authorization/Authorization.API/API/Repository/GenericRepository.cs [hunk 10]
```diff
@@ -10,15 +13,24 @@ public class GenericRepository<T>(ApplicationDbContext context) : IGenericReposi
 {
     protected readonly ApplicationDbContext Context = context ?? throw new ArgumentNullException(nameof(context));
 
-    public async Task<T> GetAsync(Guid id)
+    public async Task<T> GetAsync(object[] keys, CancellationToken cancellationToken)
```
Comment (TopSwagCode) [inline domains/authorization/Authorization.API/API/Repository/GenericRepository.cs:16]: not a big fan of object[] keys

File: domains/authorization/Authorization.API/API.IntegrationTests/API.IntegrationTests/Setup/DatabaseTest.cs (hunk 1 above)
Comment (sahma19) [reply domains/authorization/Authorization.API/API.IntegrationTests/API.IntegrationTests/Setup/DatabaseTest.cs:28]: Something sweet, and spice, and nice: https://github.com/jbogard/Respawn

File: domains/authorization/Authorization.API/API/ValueObjects/Username.cs [hunk 11]
```diff
@@ -2,15 +2,20 @@
 
 namespace API.ValueObjects;
 
-public class Name : ValueObject
+public class Username : ValueObject
 {
     public string Value { get; }
 
-    public Name(string value)
+    private Username(string value)
```
Comment (sahma19) [inline domains/authorization/Authorization.API/API/ValueObjects/Username.cs:9]: Nit: UserName rather than Username. It is a users name, and not a username for logging into a system

File: domains/authorization/Authorization.API/API/ValueObjects/ClientName.cs [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+
+        if (!IsValidClientName(value))
+            throw new ArgumentException("Value contains invalid characters.", nameof(value));
+
+        Value = value.Trim();
+    }
+
+    private static bool IsValidClientName(string value) => MyRegex().IsMatch(value);
+
+    private static Regex MyRegex() => new Regex(@"^[a-zA-Z0-9\s&.,'\-]+$");
```
Comment (sahma19) [inline domains/authorization/Authorization.API/API/ValueObjects/ClientName.cs:23]: Nit: This regex doesn't allow slashes. Can companies have slashes in their names?

File: domains/authorization/Authorization.API/API/ValueObjects/IdpUserId.cs [hunk 13]
```diff
@@ -6,11 +6,16 @@ public class IdpUserId : ValueObject
...
-    public IdpUserId(Guid value)
+    private IdpUserId(Guid value)
     {
         if (value == Guid.Empty)
             throw new ArgumentException("Value cannot be an empty Guid.", nameof(value));
 
         Value = value;
     }
+
+    public static IdpUserId Create(Guid value)
```
Comment (sahma19) [inline domains/authorization/Authorization.API/API/ValueObjects/IdpUserId.cs:17]: Can the guid be all zeroes? 0000-0000-0000-0000? Should it be allowed to?

File: domains/authorization/Authorization.API/API/ValueObjects/ClientName.cs [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+
+    public ClientName(string value)
+    {
+        if (string.IsNullOrWhiteSpace(value))
+            throw new ArgumentException("Value cannot be null, empty, or whitespace.", nameof(value));
+
+        if (!IsValidClientName(value))
+            throw new ArgumentException("Value contains invalid characters.", nameof(value));
+
+        Value = value.Trim();
```
Comment (sahma19) [inline domains/authorization/Authorization.API/API/ValueObjects/ClientName.cs:18]: What does trim do?

File: domains/authorization/Authorization.API/API/Models/Organization.cs [hunk 15]
```diff
@@ -4,47 +4,36 @@
...
     {
     }
 
     public Guid Id { get; private set; }
-    public IdpId IdpId { get; private set; } = null!;
-    public IdpOrganizationId IdpOrganizationId { get; private set; } = null!;
     public Tin Tin { get; private set; } = null!;
-    public OrganizationName OrganizationName { get; private set; } = null!;
-
+    public OrganizationName Name { get; private set; } = null!;
```
Comment (sahma19) [inline domains/authorization/Authorization.API/API/Models/Organization.cs:26]: We are using a Name Value Object now. Will this name object be used only for organizations, or for other entities? If it will be used for other entities, should we also use it for Users?

File: domains/authorization/Authorization.API/API/Models/IEntity.cs [hunk 16]
```diff
@@ -0,0 +1,6 @@
+namespace API.Models;
+
+public interface IEntity<out TId>
+{
+    TId Id { get; }
```
Comment (sahma19) [inline domains/authorization/Authorization.API/API/Models/IEntity.cs:5]: This looks spicy. What is this?

File: domains/authorization/Authorization.API/API/Models/Consent.cs [hunk 17]
```diff
@@ -32,13 +31,12 @@ private Consent()
     public Client Client { get; private set; } = null!;
     public DateTimeOffset ConsentDate { get; private set; }
 
-    public static Consent Create(Organization organization, Client client, DateTime consentDate)
+    public static Consent Create(Organization organization, Client client, DateTimeOffset consentDate)
```
Comment (sahma19) [inline domains/authorization/Authorization.API/API/Models/Consent.cs:34]: Do we use a value object for dates? I think we agreed on using longs?

File: domains/authorization/Authorization.API/API/Models/ApplicationDbContext.cs [hunk 18]
```diff
@@ -52,38 +44,39 @@ private static void ConfigureClientTable(ModelBuilder modelBuilder)
...
+        modelBuilder.Entity<User>().HasMany(it => it.Affiliations);
+
         modelBuilder.Entity<User>().HasIndex(u => u.IdpUserId).IsUnique();
     }
 
     private static void ConfigureAffiliationTable(ModelBuilder modelBuilder)
     {
         modelBuilder.Entity<Affiliation>()
-            .HasIndex(a => new { a.UserId, a.OrganizationId })
-            .IsUnique();
```
Comment (sahma19) [inline domains/authorization/Authorization.API/API/Models/ApplicationDbContext.cs:82]: Why this change?

File: domains/authorization/Authorization.API/API/Models/ApplicationDbContext.cs [hunk 19]
```diff
@@ -52,38 +44,39 @@ private static void ConfigureClientTable(ModelBuilder modelBuilder)
...
 
-        modelBuilder.Entity<User>().Property(u => u.Name)
-            .HasConversion(new ValueConverter<Name, string>(v => v.Value, v => new Name(v)))
+        modelBuilder.Entity<User>().Property(u => u.Username)
+            .HasConversion(new ValueConverter<Username, string>(v => v.Value, v => Username.Create(v)))
             .IsRequired();
 
         modelBuilder.Entity<User>().Property(u => u.IdpUserId)
-            .HasConversion(new ValueConverter<IdpUserId, Guid>(v => v.Value, v => new IdpUserId(v)))
+            .HasConversion(new ValueConverter<IdpUserId, Guid>(v => v.Value, v => IdpUserId.Create(v)))
```
Comment (sahma19) [inline domains/authorization/Authorization.API/API/Models/ApplicationDbContext.cs:61]: Why this change?

File: domains/authorization/Authorization.API/API/Models/ApplicationDbContext.cs [hunk 20]
```diff
@@ -26,23 +26,15 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
...
-        modelBuilder.Entity<Organization>().Property(o => o.IdpId)
-            .HasConversion(new ValueConverter<IdpId, Guid>(v => v.Value, v => new IdpId(v)))
-            .IsRequired();
-
-        modelBuilder.Entity<Organization>().Property(o => o.IdpOrganizationId)
-            .HasConversion(new ValueConverter<IdpOrganizationId, Guid>(v => v.Value, v => new IdpOrganizationId(v)))
-            .IsRequired();
-
-        modelBuilder.Entity<Organization>().HasIndex(o => o.IdpOrganizationId).IsUnique();
+        modelBuilder.Entity<Organization>().HasMany(it => it.Affiliations);
```
Comment (sahma19) [inline domains/authorization/Authorization.API/API/Models/ApplicationDbContext.cs:37]: Why the change here?

File: domains/authorization/Authorization.API/API/Models/Affiliation.cs [hunk 21]
```diff
@@ -6,14 +6,12 @@ namespace API.Models;
...
 {
     private Affiliation(
-        Guid id,
         Guid userId,
         Guid organizationId,
         User user,
         Organization organization
     )
     {
-        Id = id;
```
Comment (sahma19) [inline domains/authorization/Authorization.API/API/Models/Affiliation.cs:16]: No id?

File: domains/authorization/Authorization.API/API/Migrations/ApplicationDbContextModelSnapshot.cs [hunk 22]
```diff
@@ -115,9 +106,6 @@ protected override void BuildModel(ModelBuilder modelBuilder)
 
                     b.HasKey("Id");
 
-                    b.HasIndex("IdpOrganizationId")
-                        .IsUnique();
```
Comment (sahma19) [inline domains/authorization/Authorization.API/API/Migrations/ApplicationDbContextModelSnapshot.cs:119]: No index?

Also why remove uniqueness constraint?

File: domains/authorization/Authorization.API/API/Migrations/ApplicationDbContextModelSnapshot.cs [hunk 23]
```diff
@@ -50,15 +43,19 @@ protected override void BuildModel(ModelBuilder modelBuilder)
...
 
                     b.Property<string>("Name")
                         .IsRequired()
                         .HasColumnType("text");
 
-                    b.Property<int>("Role")
-                        .HasColumnType("integer");
+                    b.Property<string>("RedirectUrl")
+                        .IsRequired()
+                        .HasColumnType("text");
```
Comment (sahma19) [inline domains/authorization/Authorization.API/API/Migrations/ApplicationDbContextModelSnapshot.cs:58]: Why store as text? Shouldn't it be stored as enum(int)?

File: domains/authorization/Authorization.API/API/ValueObjects/IdpUserId.cs (hunk 13 above)
Comment (TopSwagCode) [reply domains/authorization/Authorization.API/API/ValueObjects/IdpUserId.cs:17]: As far I remember 0-0-0-0 guid is also empty guid. But not 100% sure.

File: domains/authorization/Authorization.API/API/ValueObjects/ClientName.cs (hunk 14 above)
Comment (TopSwagCode) [reply domains/authorization/Authorization.API/API/ValueObjects/ClientName.cs:18]: Removes whitespace front and back.

File: domains/authorization/Authorization.API/API.UnitTests/API.UnitTests/Repository/FakeAffilicationRepository.cs (hunk 4 above)
Comment (ckr123) [reply domains/authorization/Authorization.API/API.UnitTests/API.UnitTests/Repository/FakeAffilicationRepository.cs]: Removed repo

File: domains/authorization/Authorization.API/API.UnitTests/API.UnitTests/Repository/FakeAffilicationRepository.cs (hunk 5 above)
Comment (ckr123) [reply domains/authorization/Authorization.API/API.UnitTests/API.UnitTests/Repository/FakeAffilicationRepository.cs]: removed repo

File: domains/authorization/Authorization.API/API.UnitTests/API.UnitTests/Repository/FakeRepository.cs (hunk 6 above)
Comment (ckr123) [reply domains/authorization/Authorization.API/API.UnitTests/API.UnitTests/Repository/FakeRepository.cs]: removed

File: domains/authorization/Authorization.API/API/Authorization/[Features]/GrantConsentCommand.cs (hunk 9 above)
Comment (ckr123) [reply domains/authorization/Authorization.API/API/Authorization/[Features]/GrantConsentCommand.cs]: removed

File: domains/authorization/Authorization.API/API/Repository/GenericRepository.cs (hunk 10 above)
Comment (ckr123) [reply domains/authorization/Authorization.API/API/Repository/GenericRepository.cs]: removed

File: domains/authorization/Authorization.API/API/ValueObjects/Username.cs (hunk 11 above)
Comment (ckr123) [reply domains/authorization/Authorization.API/API/ValueObjects/Username.cs:9]: changed

File: domains/authorization/Authorization.API/API/ValueObjects/IdpUserId.cs (hunk 13 above)
Comment (ckr123) [reply domains/authorization/Authorization.API/API/ValueObjects/IdpUserId.cs:17]: both empty

File: domains/authorization/Authorization.API/API/Models/Organization.cs (hunk 15 above)
Comment (ckr123) [reply domains/authorization/Authorization.API/API/Models/Organization.cs:26]: You want one value object called Name? And be used for organisation, client and user?

File: domains/authorization/Authorization.API/API/Models/ApplicationDbContext.cs (hunk 19 above)
Comment (ckr123) [reply domains/authorization/Authorization.API/API/Models/ApplicationDbContext.cs:61]: The constructor is private

File: domains/authorization/Authorization.API/API/Models/ApplicationDbContext.cs (hunk 20 above)
Comment (ckr123) [reply domains/authorization/Authorization.API/API/Models/ApplicationDbContext.cs:37]: To generate the sql so an organization as many affiliations.

File: domains/authorization/Authorization.API/API/Models/Affiliation.cs (hunk 21 above)
Comment (ckr123) [reply domains/authorization/Authorization.API/API/Models/Affiliation.cs:16]: Id is combosite key of organization and user

File: domains/authorization/Authorization.API/API/Migrations/ApplicationDbContextModelSnapshot.cs (hunk 22 above)
Comment (ckr123) [reply domains/authorization/Authorization.API/API/Migrations/ApplicationDbContextModelSnapshot.cs:119]: The property is removed

File: domains/authorization/Authorization.API/API/Migrations/ApplicationDbContextModelSnapshot.cs (hunk 23 above)
Comment (ckr123) [reply domains/authorization/Authorization.API/API/Migrations/ApplicationDbContextModelSnapshot.cs:58]: This is the redirectUrl, it is a string

File: domains/authorization/Authorization.API/API/ValueObjects/ClientName.cs (hunk 12 above)
Comment (ckr123) [reply domains/authorization/Authorization.API/API/ValueObjects/ClientName.cs]: fixed

File: domains/authorization/Authorization.API/API/Authorization/[Features]/GetConsentForClientQuery.cs (hunk 7 above)
Comment (TopSwagCode) [reply domains/authorization/Authorization.API/API/Authorization/[Features]/GetConsentForClientQuery.cs:39]: Guids*

File: domains/authorization/Authorization.API/API.IntegrationTests/API.IntegrationTests/Setup/IntegrationTestFixture.cs [hunk 24]
```diff
@@ -0,0 +1,136 @@
...
+
+public class IntegrationTestFixture : WebApplicationFactory<Program>, IAsyncLifetime
+{
+    public PostgresContainer PostgresContainer { get; } = new();
+
+    public async Task InitializeAsync()
+    {
+        await PostgresContainer.InitializeAsync();
+
+        using var scope = Services.CreateScope();
```
Comment (martinhenningjensen) [inline domains/authorization/Authorization.API/API.IntegrationTests/API.IntegrationTests/Setup/IntegrationTestFixture.cs:37]: Delete this 37 - 40

**Review by martinhenningjensen — CHANGES_REQUESTED** [review]

File: domains/authorization/Authorization.API/API/Authorization/Controllers/ConsentController.cs [hunk 25]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        _mediator = mediator;
+        _entityDescriptor = entityDescriptor;
+    }
+
+    /// <summary>
+    /// Grants consent.
+    /// </summary>
+    [HttpPost]
+    [Authorize(Policy = "SubTypeUser")]
```
Comment (martinhenningjensen) [inline domains/authorization/Authorization.API/API/Authorization/Controllers/ConsentController.cs:31]: This policy isn't defined anywhere as far as i know. Maybe delete this and add a subtask to the story?

**Review by martinhenningjensen — CHANGES_REQUESTED** [review]

File: domains/authorization/Authorization.API/API/Authorization/Controllers/ConsentController.cs (hunk 25 above)
Comment (ckr123) [reply domains/authorization/Authorization.API/API/Authorization/Controllers/ConsentController.cs:31]: The policy is defined in program.cs
```
builder.Services.AddAuthorizationBuilder()
    .AddPolicy("SubTypeUser", policy => policy.RequireClaim("sub_type", "user")
);
```

File: domains/authorization/Authorization.API/API.IntegrationTests/API.IntegrationTests/Setup/IntegrationTestFixture.cs (hunk 24 above)
Comment (ckr123) [reply domains/authorization/Authorization.API/API.IntegrationTests/API.IntegrationTests/Setup/IntegrationTestFixture.cs]: removed

File: domains/authorization/Authorization.API/API/Authorization/Controllers/ConsentController.cs (hunk 25 above)
Comment (martinhenningjensen) [reply domains/authorization/Authorization.API/API/Authorization/Controllers/ConsentController.cs:31]: Ok, jeg tænker at den forsvinder når vi opgraderer til nyeste version af TokenValidation, så må vi lige have den tilføjet der i stedet.
